beater/beater.go

```diff
@@ -264,6 +273,73 @@ type reloader struct {
 	outputConfig agentconfig.Namespace
 	fleetConfig  *config.Fleet
 	runner       *serverRunner
+	debouncer    *debouncer
+}
+
+// debouncer wraps a function fn that is to be debounced for the length of
+// timeout. Additional calls to fire fn will reset the timer. After timeout
+// elapses, fn is fired.
+type debouncer struct {
+	triggerc chan chan<- error
+	timeout  time.Duration
+	fn       func() error
+
+	// Used to show execution of fn is currently being debounced. Must be a
+	// buffered channel with length 1.
+	active chan struct{}
+}
+
+// trigger sends a request to fire the function fn. a buffered channel which
+// will contain the return value of fn is returned to the caller, which they
+// are responsible for draining.
+func (d *debouncer) trigger(ctx context.Context) <-chan error {
+	res := make(chan error, 1)
+	select {
+	case d.active <- struct{}{}:
+		// Update the internal state to show we're currently debouncing
+		// fn execution and start the debounce method.
+		go func() {
+			d.debounce(ctx)
```

Comment:
Don't we need to send the result of debounce back to res? Otherwise the error won't be propagated and line 308 will never happen either.

Reply:
I believe that, when we send `res` on line 308, that will be added to the slice of channels that will have the error sent to them. I think this is being verified in the test.

However, maybe the code would be more readable if we send on this channel, and move `d.triggerc <- res` under `default`?